src/win32/Watcher.cpp

```diff
@@ -0,0 +1,276 @@
+  std::wstring::size_type found = path.rfind('\\');
+  std::wstringstream utf16DirectoryStream;
+
+  utf16DirectoryStream << mPath;
+
+  if (found != std::wstring::npos) {
+      utf16DirectoryStream
+        << "\\"
+        << path.substr(0, found);
+  }
+
+  std::wstring uft16DirectoryString = utf16DirectoryStream.str();
+  int utf8length = WideCharToMultiByte(
+      CP_UTF8,
+      0,
+      uft16DirectoryString.data(),
+      -1,
+      0,
+      0,
+      NULL,
+      NULL
+  );
+  char *utf8CString = new char[utf8length];
+  int failureToResolveToUTF8 = WideCharToMultiByte(
+      CP_UTF8,
+      0,
+      uft16DirectoryString.data(),
+      -1,
+      utf8CString,
+      utf8length,
+      NULL,
+      NULL
+  );
+
+  std::string utf8Directory = utf8CString;
+  delete[] utf8CString;
+
+  return utf8Directory;
+}
+
+static
+std::string getUTF8FileName(std::wstring path) {
+  std::wstring::size_type found = path.rfind('\\');
+  if (found != std::wstring::npos) {
+      path = path.substr(found + 1);
+  }
+
+  int utf8length = WideCharToMultiByte(
+      CP_UTF8,
+      0,
+      path.data(),
+      -1,
+      0,
+      0,
+      NULL,
+      NULL
+      );
+
+  // TODO: failure cases for widechar conversion
+  char *utf8CString = new char[utf8length];
+  int failureToResolveToUTF8 = WideCharToMultiByte(
+      CP_UTF8,
+      0,
+      path.data(),
+      -1,
+      utf8CString,
+      utf8length,
+      NULL,
+      NULL
+      );
+
+  std::string utf8Directory = utf8CString;
+  delete[] utf8CString;
+
+  return utf8Directory;
+}
+
+Watcher::Watcher(std::shared_ptr<EventQueue> queue, HANDLE dirHandle, const std::wstring &path)
+  : mRunning(false)
+  , mDirectoryHandle(dirHandle)
+  , mQueue(queue)
+  , mPath(path)
+{
+  ZeroMemory(&mOverlapped, sizeof(OVERLAPPED));
+  mOverlapped.hEvent = this;
+  resizeBuffers(1024 * 1024);
+  start();
+}
+
+Watcher::~Watcher()
+{
+  stop();
+}
+
+void Watcher::resizeBuffers(std::size_t size)
+{
+  mReadBuffer.resize(size);
+  mWriteBuffer.resize(size);
+}
+
+void Watcher::run()
+{
+  while(mRunning) {
+    SleepEx(INFINITE, true);
+  }
+}
+
+bool Watcher::loop()
+{
+  DWORD bytes = 0;
+
+  if (!isRunning()) {
+    return false;
+  }
+
+  if (!ReadDirectoryChangesW(
+    mDirectoryHandle,
+    mWriteBuffer.data(),
+    static_cast<DWORD>(mWriteBuffer.size()),
+    TRUE,                           //recursive watching
+    FILE_NOTIFY_CHANGE_FILE_NAME
+    | FILE_NOTIFY_CHANGE_DIR_NAME
+    | FILE_NOTIFY_CHANGE_ATTRIBUTES
+    | FILE_NOTIFY_CHANGE_SIZE
+    | FILE_NOTIFY_CHANGE_LAST_WRITE
+    | FILE_NOTIFY_CHANGE_LAST_ACCESS
+    | FILE_NOTIFY_CHANGE_CREATION
+    | FILE_NOTIFY_CHANGE_SECURITY,
+    &bytes,                         //num bytes written
+    &mOverlapped,
+    [](DWORD errorCode, DWORD numBytes, LPOVERLAPPED overlapped) {
+      auto watcher = reinterpret_cast<Watcher*>(overlapped->hEvent);
+      watcher->eventCallback(errorCode);
+    }))
+  {
```

Comment:
{ on the same line, It also looks like 148 may have too many spaces for the close (it should line up with the `if` on 130.